jsk_pcl_ros/src/euclidean_cluster_extraction_nodelet.cpp

```diff
@@ -42,109 +42,269 @@ using namespace pcl;
+        EuclideanClusterExtraction<pcl::PointXYZ> impl;
+        if (filtered_cloud->points.size() > 0) {
+          jsk_topic_tools::ScopedTimer timer = kdtree_acc_.scopedTimer();
 #if ( PCL_MAJOR_VERSION >= 1 && PCL_MINOR_VERSION >= 5 )
-      pcl::search::KdTree<pcl::PointXYZ>::Ptr tree (new pcl::search::KdTree<pcl::PointXYZ>);
-      tree = boost::make_shared< pcl::search::KdTree<pcl::PointXYZ> > ();
+          pcl::search::KdTree<pcl::PointXYZ>::Ptr tree (new pcl::search::KdTree<pcl::PointXYZ>);
+          tree = boost::make_shared< pcl::search::KdTree<pcl::PointXYZ> > ();
 #else
-      pcl::KdTree<pcl::PointXYZ>::Ptr tree (new pcl::KdTreeFLANN<pcl::PointXYZ>);
-      tree = boost::make_shared<pcl::KdTreeFLANN<pcl::PointXYZ> > ();
+          pcl::KdTree<pcl::PointXYZ>::Ptr tree (new pcl::KdTreeFLANN<pcl::PointXYZ>);
+          tree = boost::make_shared<pcl::KdTreeFLANN<pcl::PointXYZ> > ();
 #endif
-      tree->setInputCloud (cloud);
-      impl.setClusterTolerance (tolerance);
-      impl.setMinClusterSize (minsize_);
-      impl.setMaxClusterSize (maxsize_);
-      impl.setSearchMethod (tree);
-      impl.setIndices(nonnan_indices);
-      impl.setInputCloud (cloud);
+          tree->setInputCloud(filtered_cloud);
+          impl.setClusterTolerance(tolerance);
+          impl.setMinClusterSize(minsize_);
+          impl.setMaxClusterSize(maxsize_);
+          impl.setSearchMethod(tree);
+          impl.setInputCloud(filtered_cloud);
+        }
+
+        std::vector<pcl::PointIndices> output_indices;
+        {
+          jsk_topic_tools::ScopedTimer timer = segmentation_acc_.scopedTimer();
+          impl.extract(output_indices);
+          if (downsample_) {
+            for (size_t i = 0; i < output_indices.size(); ++i) {
+              pcl::PointIndices::Ptr tmp_indices(new pcl::PointIndices);
+              for (size_t j = 0; j < output_indices.at(i).indices.size(); ++j) {
+                tmp_indices->indices.insert(
+                  tmp_indices->indices.end(),
+                  downsample_to_original_indices_[nonnan_indices->indices[output_indices.at(i).indices[j]]].begin(),
+                  downsample_to_original_indices_[nonnan_indices->indices[output_indices.at(i).indices[j]]].end());
+              }
+              removeDuplicatedIndices(tmp_indices);
+              output_indices[i] = *tmp_indices;
+            }
+          } else {
+            for (size_t i = 0; i < output_indices.size(); ++i) {
+              pcl::PointIndices tmp_indices;
+              tmp_indices.indices.resize( output_indices.at(i).indices.size());
+              for (size_t j = 0; j < output_indices.at(i).indices.size(); ++j) {
+                tmp_indices.indices[j] = nonnan_indices->indices[output_indices.at(i).indices[j]];
+              }
+              output_indices[i] = tmp_indices;
+            }
+          }
+        }
+        std::vector<int> result_indices;
+        std::vector<int> examine_indices;
+        if (keep_whole_cluster) {
```

Comment:
it is better to add `keep_whole_cluster` or `cluster_filter: all` `cluster_filter: max_size` in `rosparam`

Reply:
Nice suggestion!
I add `cluster_filter` enum type.